File: myFinance_improved_V2/lib/features/inventory_analysis/supply_chain/domain/entities/supply_chain_status.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'supply_chain_status.freezed.dart';

/// Risk level enum
enum RiskLevel {
  critical,
  warning,
  normal,
}

/// Supply chain problem product data
/// Error Integral = shortage_days × avg_shortage_per_day
@freezed
class SupplyChainProduct with _$SupplyChainProduct {
  const SupplyChainProduct._();

  const factory SupplyChainProduct({
    required String productId,
    required String productName,
    required int shortageDays,
    required num avgShortagePerDay,
    required num totalShortage,
    required num errorIntegral,
    required String riskLevel, // 'Critical', 'Warning', 'Normal'
  }) = _SupplyChainProduct;

  RiskLevel get riskLevelType {
    return switch (riskLevel) {
      'Critical' => RiskLevel.critical,
      'Warning' => RiskLevel.warning,
      _ => RiskLevel.normal,
    };
  }

  /// Risk label
  String get riskLabel {
    return switch (riskLevelType) {
      RiskLevel.critical => 'Critical',
      RiskLevel.warning => 'Warning',
      RiskLevel.normal => 'Normal',
    };
  }

  /// Error Integral description
  String get integralDescription =>
      '${shortageDays} days × avg ${avgShortagePerDay.toStringAsFixed(1)} shortage';
}

/// Supply chain status aggregate data
/// RPC: get_supply_chain_status response
@freezed
class SupplyChainStatus with _$SupplyChainStatus {
  const SupplyChainStatus._();

  const factory SupplyChainStatus({
    required List<SupplyChainProduct> urgentProducts,
  }) = _SupplyChainStatus;

  /// Critical product count
  int get criticalCount =>
      urgentProducts.where((p) => p.riskLevelType == RiskLevel.critical).length;

  /// Warning product count
  int get warningCount =>
      urgentProducts.where((p) => p.riskLevelType == RiskLevel.warning).length;

  /// Overall status
  String get status {
    if (criticalCount > 0) return 'critical';
    if (warningCount > 0) return 'warning';
    return 'good';
  }

  /// Whether data is empty
  bool get isEmpty => urgentProducts.isEmpty;

  /// Status text
  String get statusText {
    if (isEmpty) return 'No risk products';
    if (criticalCount > 0) return '$criticalCount critical';
    if (warningCount > 0) return '$warningCount warning';
    return 'Normal';
  }
}

File: website/src/features/inventory-analysis/domain/entities/supplyChainStatus.ts
/**
 * Supply Chain Status Entity
 * Domain entity for RPC response mapping
 */

// Urgent product with supply chain risk
export interface UrgentSupplyChainProduct {
  productName: string;
  delayDays: number;
  avgShortage: number;
  totalShortage: number;
  errorIntegral: number;
  riskLevel: 'Critical' | 'Warning';
}

// Main supply chain status entity
export interface SupplyChainStatus {
  urgentProducts: UrgentSupplyChainProduct[];
}

/**
 * Maps RPC response to domain entity
 */
export function mapSupplyChainStatusFromRpc(data: Record<string, unknown>): SupplyChainStatus {
  const urgentProductsData = data.urgent_products as Array<Record<string, unknown>> | null | undefined;

  return {
    urgentProducts: (urgentProductsData ?? []).map((prod) => ({
      productName: String(prod.product_name ?? ''),
      delayDays: Number(prod.delay_days ?? 0),
      avgShortage: Number(prod.avg_shortage ?? 0),
      totalShortage: Number(prod.total_shortage ?? 0),
      errorIntegral: Number(prod.error_integral ?? 0),
      riskLevel: (prod.risk_level === 'Critical' ? 'Critical' : 'Warning') as 'Critical' | 'Warning',
    })),
  };
}
